File: programmers/level3/단어변환.py
from collections import deque


# b를 리턴하면 변환 성공
def findWord(a, b):
    diff = 0
    for i in range(len(a)):
        if a[i] != b[i]:
            diff += 1
            if diff >= 2:
                return a
    if diff == 1:
        return b
    else:
        return a


def solution(begin, target, words):
    if target not in words:
        return 0
    answer = len(words)
    v = [False] * len(words)
    q = deque()

    for i, w in enumerate(words):
        if w == begin:
            v[i] = True

    q.append([begin, v, 1])

    while q:
        curWord, curV, curCnt = q.popleft()
        if curWord == target:
            print(curCnt)
            answer = min(answer, curCnt - 1)
        for i, w in enumerate(words):
            result = findWord(curWord, w)
            if not curV[i] and result == w:
                curV[i] = True
                q.append([result, curV, curCnt + 1])
    return answer

File: programmers/level3/단어변환.js
function solution(begin, target, words) {
  var answer = 0;
  const v = new Array(words.length).fill(false);

  q = [];
  q.push([begin, 0]);
  while (q.length > 0) {
    const [curWord, cnt] = q.shift();
    if (curWord === target) {
      answer = cnt;
      break;
    }
    for (let i = 0; i < words.length; i++) {
      if (!v[i] && checkWord(curWord, words[i])) {
        v[i] = true;
        q.push([words[i], cnt + 1]);
      }
    }
  }

  return answer;
}

function checkWord(input, compare) {
  let diff = 0;
  for (let i = 0; i < input.length; i++) {
    if (input[i] !== compare[i]) {
      diff += 1;
    }
    if (diff > 1) {
      return false;
    }
  }
  return diff === 1 ? true : false;
}
